File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/adapter/MongoStagePerformanceRollupRepository.java
package com.study.webflux.rag.infrastructure.monitoring.adapter;

import java.time.Instant;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Repository;

import com.study.webflux.rag.domain.monitoring.entity.StagePerformanceRollupEntity;
import com.study.webflux.rag.domain.monitoring.model.MetricsGranularity;
import com.study.webflux.rag.domain.monitoring.model.StagePerformanceRollup;
import com.study.webflux.rag.domain.monitoring.port.StagePerformanceRollupRepository;
import com.study.webflux.rag.infrastructure.monitoring.repository.SpringDataStagePerformanceRollupRepository;
import reactor.core.publisher.Flux;

@Repository
@RequiredArgsConstructor
public class MongoStagePerformanceRollupRepository implements StagePerformanceRollupRepository {

	private final SpringDataStagePerformanceRollupRepository repository;

	@Override
	public Flux<StagePerformanceRollup> saveAll(Flux<StagePerformanceRollup> rollups) {
		return repository.saveAll(rollups.map(StagePerformanceRollupEntity::fromDomain))
			.map(StagePerformanceRollupEntity::toDomain);
	}

	@Override
	public Flux<StagePerformanceRollup> findByGranularityAndBucketStartBetween(
		MetricsGranularity granularity,
		Instant startTime,
		Instant endTime) {
		return repository.findByGranularityAndBucketStartBetween(
			granularity.name(),
			startTime,
			endTime)
			.map(StagePerformanceRollupEntity::toDomain);
	}
}

File: monitoring/src/main/kotlin/com/miyou/app/monitoring/adapter/MongoStagePerformanceRollupRepository.kt
package com.miyou.app.monitoring.adapter

import com.miyou.app.monitoring.model.MetricsGranularity
import com.miyou.app.monitoring.model.StagePerformanceRollup
import com.miyou.app.monitoring.port.StagePerformanceRollupRepository
import com.miyou.app.monitoring.document.StagePerformanceRollupDocument
import com.miyou.app.monitoring.repository.SpringDataStagePerformanceRollupRepository
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import java.time.Instant

@Repository
class MongoStagePerformanceRollupRepository(
    private val repository: SpringDataStagePerformanceRollupRepository,
) : StagePerformanceRollupRepository {
    override fun saveAll(rollups: Flux<StagePerformanceRollup>): Flux<StagePerformanceRollup> =
        repository
            .saveAll(rollups.map(StagePerformanceRollupDocument::fromDomain))
            .map(StagePerformanceRollupDocument::toDomain)

    override fun findByGranularityAndBucketStartBetween(
        granularity: MetricsGranularity,
        startTime: Instant,
        endTime: Instant,
    ): Flux<StagePerformanceRollup> =
        repository
            .findByGranularityAndBucketStartBetween(
                granularity.name,
                startTime,
                endTime,
            ).map(StagePerformanceRollupDocument::toDomain)
}
